### PR TITLE
Create Pip Package

### PR DESCRIPTION
**Pull request by abetlen**
Hey @saharNooby here is some preliminary work to set up a python package similar to the one I've set up in llama-cpp-python using scikit-build to handle building the shared library. This is still a work-in-progress so I'm totally open to change anything / answer any questions / etc. Let me know what you think

Changes
- Add scikit-build artifacts to gitignore
- Add shared library target to Makefile
- Add scikit-build specific section to CMakeLists
- Add project scaffolding files (setup.py, pyproject.toml)
- Update shared library name for macos (use Makefile instead of cmake to avoid M1 architecture issues)
- Search for shared object file in package
- I also created a package on PyPI just to test this out, if you have a username there I can add you as the project owner :)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -245,3 +245,28 @@ if (BUILD_SHARED_LIBS)
     set_target_properties(rwkv PROPERTIES POSITION_INDEPENDENT_CODE ON)
     target_compile_definitions(rwkv PRIVATE RWKV_SHARED RWKV_BUILD)
 endif()
+
+if(SKBUILD)
```
Comment (saharNooby) [inline CMakeLists.txt:249]:
```suggestion
if (SKBUILD)
```

File: Makefile [hunk 2]
```diff
@@ -227,3 +227,6 @@ ggml.o: ggml.c ggml.h
 
 rwkv.o: rwkv.cpp rwkv.h
 	$(CXX) $(CXXFLAGS) -c rwkv.cpp -o rwkv.o
+
+librwkv.so: rwkv.o ggml.o
```
Comment (saharNooby) [inline Makefile:231]: Will it work when building on Windows?

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+from skbuild import setup
+
+setup(
+    name="rwkv_cpp_python",
```
Comment (saharNooby) [inline setup.py:4]: I guess just `rwkv_cpp` or more preferred `rwkv-cpp` is fine, since this is already a Python package :)

**Comment by saharNooby** [comment]
Hure thanks! Sorry if it is inappropriate to leave comments on draft PR. Let me know when this is ready for me to check whether it builds!

BTW, do I need to register somewhere/do something extra to publish the package to `pip`?

File: Makefile (hunk 2 above)
Comment (abetlen) [reply Makefile:231]: I should note that when buliding through python via scikit-build the user doesn't have to do anything special, inside the `SKBUILD` if block in the root CMakeLists.txt the build decides whether to build using cmake or just directly invoke the existing Makefile based on the OS. The `SKBUILD` check should also ensure that the CMakeLists.txt should still work if you just want to build the project as a library but don't care about the python package.

File: setup.py (hunk 3 above)
Comment (abetlen) [reply setup.py:4]: Sounds good, I'll update this as well. I believe both alias to the same name on PyPI.

**Comment by abetlen** [comment]
> Hure thanks! Sorry if it is inappropriate to leave comments on draft PR. Let me know when this is ready for me to check whether it builds!
> 
> BTW, do I need to register somewhere/do something extra to publish the package to `pip`?

Yes, just head over to https://pypi.org/account/register/ and once you've registered let me know the username.

**Comment by saharNooby** [comment]
@abetlen Registered on PyPi: https://pypi.org/user/saharNooby/

**Comment by abetlen** [comment]
> @abetlen Registered on PyPi: https://pypi.org/user/saharNooby/

Awesome, invite sent.

**Comment by abetlen** [comment]
Latest changes:

- Had to rename `rkwv` folder to `rwkv_cpp`, unfortunately no way around this as python package import __must__ have the same name as the folder they sit in and there's already an official non-cpu RWKV package so this avoids conflicts.
- Removed the nested requirements file as this is already covered by the setup.py script

**Comment by iacore** [comment]
Is this branch stable? I'm going to back-port my changes onto this.

**Comment by abetlen** [comment]
@iacore yes these are all my changes so far, really appreciate that! Sorry I meant to put more work in here but I got a little busy in the last week.

**Pull request closed without merging**